### ISSUE TITLE
3D plots for graphs

### ISSUE THREAD

**Issue by sadamov** [mllam/weather-model-graphs#20]
**Title:** Add 3D Graph Plotting Functionality Using Plotly

**Description:**
We should add functionality to visualize the graph structures in 3D using Plotly. This will help in better understanding the spatial relationships between nodes and edges in our graph-based neural weather prediction models.

**Tasks:**
1. Implement a function to plot the graph structure in 3D using Plotly.
2. The function should support both hierarchical and non-hierarchical graphs.
3. Add options to save the plot as an interactive HTML file and to toggle axis visibility.
4. Ensure that the function can handle different types of edges (e.g., M2G, G2M, M2M) with distinct colors and line widths.
5. Integrate the function into the existing codebase, ensuring compatibility with the current graph data structures.

Example:
https://github.com/mllam/neural-lam/blob/5b71be3c68d815e0e376ee651c14f09d801f86de/plot_graph.py

**Challenges:**
The html files can easily be >100MB in size, crashing most browsers. We might need to implement some performance/size tricks to prevent this.

**Surplus:**
3D objects could be rendered for Blender, which supports a Python-like DSL as a backend. This would allows us to benefit from all the 3D modeling techniques (e.g. shaders) for even nicer looking graphs.

**Comment by leifdenby** [comment]
> The html files can easily be >100MB in size, crashing most browsers. We might need to implement some performance/size tricks to prevent this.

Would this also be the case if using [panel](https://panel.holoviz.org/) to create a python application within which the serialisation to allow plotly to do the rendering is handle by panel? It might be that they have worked out some smart way of doing this :) I.e. I am suggesting to not generate static html, but instead create a visualisation application which can be started with for example:

``` bash
python -m weather_model_graphs.viz --interactive graph_file.pickle
```

obviously this require writing the graph to some serialised format which we can't currently do, but we could start by pickling the `nx.Graph` object. That might be an easy starting point? :)

**Comment by joeloskarsson** [comment]
A note about this challenge: The standard behavior in plotly, and used in the neural-lam script linked, is to set up a web server that serves an interactive plotting application. I do however not know if this just creates a static html somewhere and serves that or it is doing something smarter. I just know that I have never had any issues with opening these interactive plots in a browser, but perhaps that could be the case if you have a very large amount of nodes/edges. You can also save the whole thing to an html file, and those can indeed become very large.

**Comment by sadamov** [comment]
I wonder if this might be useful here: https://hvplot.holoviz.org/

**Comment by joeloskarsson** [comment]
It would be really nice to also actually add the earth (textured) to the 3d plot, rather than just a bunch of grid nodes. Going by this post https://community.plotly.com/t/applying-full-color-image-texture-to-create-an-interactive-earth-globe/60166 it looks fairly simple :smile:

**Comment by sadamov** [comment]
Posting another library that might be useful here: https://github.com/benbovy/spherely

**Comment by joeloskarsson** [comment]
Ooh, nice! This might actually be super useful for LAM graph creation in lat-lons. Me and @TomasLandelius discussed a bit back on how to implement convex hulls on the surface of a sphere and check which points lay inside it. This might be straightforward with something like https://spherely.readthedocs.io/en/latest/_api_generated/spherely.convex_hull.html#spherely.convex_hull + https://spherely.readthedocs.io/en/latest/_api_generated/spherely.contains.html#spherely.contains .

**Comment by TomasLandelius** [comment]
I did it with

from spherical_geometry.polygon import SphericalPolygon
...
chull_lam = SphericalPolygon.convex_hull(pts_lam_x)
msk_in = np.array([chull_lam.contains_lonlat(x[1], x[0]) for x in points_bnd])

but the spherely package should be fine as well I think
________________________________________

**Comment by AdMub** [comment]
Hi everyone! 👋 
I was looking into this issue and put together a lightweight Proof of Concept (PoC) using `plotly` to visualize the hierarchical mesh in 3D.

To address the performance/browser-crashing concerns mentioned above (the >100MB HTML files), I batched the edge coordinates into a single `go.Scatter3d` trace separated by `None`. This keeps the rendering blazing fast and the HTML file very small, even with thousands of edges. I also `mapped` the level attribute to the Z-axis to cleanly separate the hierarchy.

(Here is a screenshot of the interactive output:)

<img width="1822" height="947" alt="Image" src="https://github.com/user-attachments/assets/6f256dba-43ca-455d-b9ad-edae2237d5d0" />

Before I open a PR to add this as `plot_3d.py` alongside the existing `plot_2d.py` in the `visualise` directory, I wanted to get your thoughts on the recent discussion regarding the "Earth sphere" projection.

Would you prefer to merge a fast, flat Cartesian 3D plotter like this first as a baseline tool, or would you prefer I jump straight into projecting these nodes onto a spherical/lat-lon globe (perhaps exploring the `spherely` integration @joeloskarsson and Tomas mentioned)?

Happy to take this in whichever direction is most useful for the team right now!

**Comment by AdMub** [comment]
Hi @joeloskarsson, that is a fantastic architectural point. We definitely don't want to duplicate rendering logic across repositories or lock neural-lam users out of visualization tools.

Since `neural-lam` already lists `weather-model-graphs` as a dependency (as of PR #164), what if we structure it like this?
1. We build the core `plot_3d()` function here in `weather-model-graphs` inside a dedicated `mllam_graphs.visualize` module.
2. The function is designed to accept standard NetworkX `DiGraph` objects (with positional attributes), meaning it is completely agnostic to how the graph was generated.
3. Over in `neural-lam`, instead of maintaining its own `plot_graph.py` script, it simply imports and wraps this centralized function (e.g., `from mllam_graphs.visualize import plot_3d`).

This keeps the heavy lifting (Plotly batching, performance optimizations) centralized here in WMG where the graph structures are defined, while allowing anyone using `neural-lam` with a custom graph to seamlessly pass it into the visualizer.

Does that align with how you and @leifdenby envision the boundary between the two packages?

**Comment by leifdenby** [comment]
> @joeloskarsson  In one way it makes sense that wmg has graph-visualization, but it also seems that there is large interest in having neural-lam working also with graphs not coming directly from wmg. In that case, if all vis sit in this repo we could not plot these.
> Thoughts [@leifdenby](https://github.com/leifdenby) ?

> @Diya910: If the visualisation logic is centralized in WMG but designed to operate purely on a generic networkx.Graph (with minimal assumptions like positional attributes), that would keep it reusable while still allowing Neural-LAM, including custom graph users to benefit from it.

Yes, this was also my thought. To make general purpose visualisation implementation anywhere (as-in in any software package) we either need to agree on 1) a disk-format or 2) a python-object representation. I am still leaning in the direction that `networkx.DiGraph` is a useful representation to start from, because the disk formats are quite diverse at the moment (and evolving). I think having visualisation in `wmg` makes sense to me because visualising the graph is already useful there before it is consumed in say `neural-lam`. However, I do see the merit in being able to take for example latent representations and visualise them on the graph in a downstream application in `neural-lam`. To facilitate that I would suggest that we need the functionality to take information derived in `neural-lam` and inject it back into the graph by setting new attributes on an existing graph from `wmg`.

> @sadamov: Just posting some inspiration here, from a colleague at MeteoSwiss who vizualised anemoi-graphs in a prototype:

Yes, that is a really nice visualisation! I saw the example, but I haven't looked at the code. Do you know what data-structure is being used for this? My feeling this taking the route of `anemoi-graphs` disks format -> loaded into `HeteroData` object -> serialised to json inside the HTML file (looks like using a jinja2 template) -> rendered in opengl. I think a better approach might be to 1) split frontend doing the opengl rendering from a backend (e.g. in flask) serving the data, 2) abstract the backend so it can be instantiated either from the `networkx.DiGraph` object or a `HeteroData` object. That way it could be used from either representation. Do you have a link to the code for the implementation of that example @sadamov?

> but it also seems that there is large interest in having neural-lam working also with graphs not coming directly from wmg

Ok, could you link me to an issue on this @joeloskarsson? In general I think it would be really good idea to allow for different ways of creating graphs for `neural-lam`. I think a key step in that direction is formalising the specification for the current pytorch-tensor format that we use. I have been working on a validator and spec for that, but I'm not quite done. I can try and finish the first draft of that next week, unless you are thinking something else is the right approach here?
